snapcraft/meta/snap_yaml.py

```diff
@@ -259,52 +244,34 @@ class SnapMetadata(SnapcraftMetadata):
     """
 
     name: str
-    title: Optional[str]
+    title: str | None = None
     version: str
     summary: SummaryStr
     description: str
-    license: Optional[str]
-    type: Optional[str]
-    architectures: List[str]
-    base: Optional[str]
-    assumes: Optional[List[str]]
-    epoch: Optional[str]
-    apps: Optional[Dict[str, SnapApp]]
+    license: str | None = None
+    type: str | None = None
+    architectures: list[str]
+    base: str | None = None
+    assumes: list[str] | None = None
+    epoch: str | None = None
+    apps: dict[str, SnapApp] | None = None
     confinement: str
     grade: str
-    environment: Optional[Dict[str, Any]]
-    plugs: Optional[Dict[str, Any]]
-    slots: Optional[Dict[str, Any]]
-    hooks: Optional[Dict[str, Any]]
-    layout: Optional[
-        Dict[str, Dict[Literal["symlink", "bind", "bind-file", "type"], str]]
-    ]
-    system_usernames: Optional[Dict[str, Any]]
-    provenance: Optional[str]
-    links: Optional[Links]
-    components: Optional[Dict[str, ComponentMetadata]]
-
-    @classmethod
-    def unmarshal(cls, data: Dict[str, Any]) -> "SnapMetadata":
-        """Create and populate a new ``SnapMetadata`` object from dictionary data.
-
-        The unmarshal method validates entries in the input dictionary, populating
-        the corresponding fields in the data object.
-
-        :param data: The dictionary data to unmarshal.
-
-        :return: The newly created object.
-
-        :raise TypeError: If data is not a dictionary.
-        """
-        if not isinstance(data, dict):
-            raise TypeError("data is not a dictionary")
-
-        return cls(**data)
-
-    def get_provider_content_directories(self) -> List[Path]:
+    environment: dict[str, Any] | None = None
+    plugs: dict[str, Any] | None = None
+    slots: dict[str, Any] | None = None
+    hooks: dict[str, Any] | None = None
+    layout: (
+        dict[str, dict[Literal["symlink", "bind", "bind-file", "type"], str]] | None
```

Comment:
Each layout entry's value is a `SingleEntryDict`, isn't it?

```suggestion
        dict[str, SingleEntryDict[Literal["symlink", "bind", "bind-file", "type"], str]] | None
```